language/move-compiler/src/hlir/translate.rs

```diff
@@ -1187,9 +1216,20 @@ fn exp_impl(
                 parameter_types,
                 acquires,
             } = *call;
-            let expected_type = H::Type_::from_vec(eloc, single_types(context, parameter_types));
+            let expected_single_types = single_types(context, parameter_types);
+            let expected_type = H::Type_::from_vec(eloc, expected_single_types.clone());
             let htys = base_types(context, type_arguments);
             let harg = exp(context, result, Some(&expected_type), *arguments);
+            for (i, ty) in expected_single_types.into_iter().enumerate() {
+                if ty.value.is_mut_ref() {
+                    match harg.exp_at_index(i).exp.value {
```

Comment:
TooManyArgs is not a blocking error, so these might not line up. (I think this `harg.exp_at_index(i)` might fail).